### ISSUE TITLE
[Question] How to use FTP upload?

### ISSUE THREAD

**Issue by zentoaku** [jenkinsci/azure-app-service-plugin#32]
Hi,

I've set up the connection to Azure Web App using Post-build action "Publish an Azure Web App". The problem I see is that it tries to  connect to some sort of GIT repository there, which I'm not utilizing (at least to my knowledge).
So the question is: how to use FTP instead of GIT? Is it possible only through Jenkins Pipeline?

Sorry if my questions feel dumb but I'm completely new to Azure and I'm discovering everything along the way :)

**Comment by ArchangelSDY** [comment]
FTP is deprecated in latest version. Now [WAR Deploy](https://docs.microsoft.com/en-us/azure/app-service/app-service-deploy-zip#deploy-war-file) is used for Java app, and [Git Deploy](https://docs.microsoft.com/en-us/azure/app-service/app-service-deploy-local-git) for other languages.

**Comment by zentoaku** [comment]
Oh, that might be an issue. So what should I do in the following scenario:
- local GIT,
- local Jenkins,
- Build made on Jenkins,
- Want to just upload compiled sources to Azure Web app?

**Comment by ArchangelSDY** [comment]
Azure Web App just use Git as something like storage rather than source control. You can specify your build artifacts and the plugin will check in and push those files to server.

**Comment by zentoaku** [comment]
Is it possible to use Jenkins build result/workspace files?
I don't get what GIT is for in this flow. What I imagine is:
private git -> jenkins pull -> jenkins build -> upload dist folder to Azure Web Apps

Can you shed some light on this? Am I completely wrong with my assumptions and what I seek in comparison to what this plugin and Azure allows?

**Comment by ArchangelSDY** [comment]
This plugin does the job that **upload dist folder to Azure Web App** in your flow. Git is one of the two ways to do that. Each Azure Web App has a corresponding Git server. This plugin checks in your **dist files** and does git push to the server. Azure Web App then automatically triggers a git pull to deploy your files to target location. So here Git is just a way to upload your files. Nothing related to your source code.

You can find details on how this works in [Kudu documents](https://github.com/projectkudu/kudu/wiki/Deployment).

In short, just fill in your dist files pattern in `filePath` field and don't worry about Git. That's internal implementation details. If you met some issues when use Git via http proxy, see [this article](https://mathewsachin.github.io/blog/2016/12/29/git-auth-proxy.html).

**Comment by ArchangelSDY** [comment]
BTW, FTP deploy is still in code base. If you prefer that you can change this line:
https://github.com/jenkinsci/azure-app-service-plugin/blob/9882789dc1a598430165b47aee12817ea982ea82/src/main/java/com/microsoft/jenkins/appservice/WebAppDeploymentCommandContext.java#L153
to
```java
 builder.withStartCommand(FTPDeployCommand.class); 
```

**Comment by zentoaku** [comment]
Awesome :) Thanks for the explanation, I will try that and get back to you :D

**Comment by zentoaku** [comment]
Turned out that I've made mistake with setting up source folder and the plugin wasn't throwing any error if the path with sources doesn't exist.

After fixing it I see some progress but new error appears:
```
> git push https://xxx.scm.azurewebsites.net:443/xxx.git master:master
ERROR: Build step failed with exception
hudson.plugins.git.GitException: Command "git push https://xxx.scm.azurewebsites.net:443/xxx.git master:master" returned status code 128:
stdout: 
stderr: fatal: unable to access 'https://xxx.scm.azurewebsites.net:443/xxx.git/': Invalid file descriptor

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:2002)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1721)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$300(CliGitAPIImpl.java:72)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$8.execute(CliGitAPIImpl.java:2081)
	at com.microsoft.jenkins.appservice.commands.GitDeployCommand.execute(GitDeployCommand.java:114)
	at com.microsoft.jenkins.appservice.commands.GitDeployCommand.execute(GitDeployCommand.java:1)
	at com.microsoft.jenkins.azurecommons.command.CommandService.runCommand(CommandService.java:88)
	at com.microsoft.jenkins.azurecommons.command.CommandService.execute(CommandService.java:96)
	at com.microsoft.jenkins.azurecommons.command.CommandService.executeCommands(CommandService.java:75)
	at com.microsoft.jenkins.azurecommons.command.BaseCommandContext.executeCommands(BaseCommandContext.java:77)
	at com.microsoft.jenkins.appservice.WebAppDeploymentRecorder.perform(WebAppDeploymentRecorder.java:203)
	at hudson.tasks.BuildStepCompatibilityLayer.perform(BuildStepCompatibilityLayer.java:81)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:744)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:690)
	at hudson.model.Build$BuildExecution.post2(Build.java:186)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:635)
	at hudson.model.Run.execute(Run.java:1823)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:429)
Build step 'Publish an Azure Web App' marked build as failure
Finished: FAILURE
```
Any idea what might be happening here?

**Comment by zentoaku** [comment]
I've increased git http buffer `git config http.postBuffer 524288000` and it solved the errror.

Build and push to Azure now works without error but I don't see any changes online. Do I need some additional setup on Azure do make it work?
```
Seen branch in repository origin/master
Seen 1 remote branch
 > git config core.sparsecheckout # timeout=10
 > git checkout -f master
 > git add angular/dist/0.000bc02b7f8fef843f62.chunk.js # timeout=10
 > git add angular/dist/3rdpartylicenses.txt # timeout=10
 > git add angular/dist/MaterialIcons-Regular.012cf6a10129e2275d79.woff # timeout=10
 > git add angular/dist/MaterialIcons-Regular.570eb83859dc23dd0eec.woff2 # timeout=10
 > git add angular/dist/MaterialIcons-Regular.a37b0c01c0baf1888ca8.ttf # timeout=10
 > git add angular/dist/MaterialIcons-Regular.e79bfd88537def476913.eot # timeout=10
 > git add angular/dist/assets/Home_ContactsCoaching.svg # timeout=10
 > git add angular/dist/assets/Home_GetStarted.svg # timeout=10
 > git add angular/dist/assets/Home_ToolsTemplates.svg # timeout=10
 > git add angular/dist/assets/TopNav_UserProfile.svg # timeout=10
 > git add angular/dist/assets/UXLab_toolsAndTemplates.png # timeout=10
 > git add angular/dist/assets/UXlab_contacts.png # timeout=10
 > git add angular/dist/assets/UXlab_getStarted.png # timeout=10
 > git add angular/dist/assets/wow_icon_contactsCoaching_1920px.png # timeout=10
 > git add angular/dist/assets/wow_icon_getstarted_1920x.png # timeout=10
 > git add angular/dist/assets/wow_icon_toolsTemplates_1920px.png # timeout=10
 > git add angular/dist/favicon.ico # timeout=10
 > git add angular/dist/index.html # timeout=10
 > git add angular/dist/inline.d9b134101eb7ece0451f.bundle.js # timeout=10
 > git add angular/dist/main.e3d74b36e9a29553554d.bundle.js # timeout=10
 > git add angular/dist/polyfills.8d911633195c2fb7185f.bundle.js # timeout=10
 > git add angular/dist/styles.2fb50bd5603d6919d036.bundle.css # timeout=10
 > git add angular/dist/vendor.d283d5869414d5cccdef.bundle.js # timeout=10
 > git add angular/dist/web.config # timeout=10
Deploy repository is up-to-date. Nothing to commit.
Done Azure Web App deployment.
Finished: SUCCESS
```

**Comment by zentoaku** [comment]
I've played around with plugin configuration and it worked :D

I had to set it up like this:
```
Files: **/*
Source Directory: angular/dist
Target Directory left empty
```

Thanks for everything, it really helped me wrap my head around how Azure and this plugin works :)

**Comment by ArchangelSDY** [comment]
Great to know that!
